persist/fs/commitlog/reader.go

```diff
@@ -109,58 +165,171 @@ func (r *reader) Read() (
+	}
+}
 
-		id := r.bytesPool.Get(len(decoded.ID))
-		id.IncRef()
-		defer id.DecRef()
-		id.AppendAll(decoded.ID)
-
-		namespace := r.bytesPool.Get(len(decoded.Namespace))
-		namespace.IncRef()
-		defer namespace.DecRef()
-		namespace.AppendAll(decoded.Namespace)
-
-		metadata = Series{
-			UniqueIndex: entry.Index,
-			ID:          ts.BinaryID(id),
-			Namespace:   ts.BinaryID(namespace),
-			Shard:       decoded.Shard,
+func (r *reader) decoderLoop(inBuf <-chan decoderArg, outBuf chan<- readResponse) {
+	decodingOpts := r.opts.FilesystemOptions().DecodingOptions()
+	decoder := msgpack.NewDecoder(decodingOpts)
+	metadataDecoder := msgpack.NewDecoder(decodingOpts)
+
+	for arg := range inBuf {
+		readResponse := readResponse{}
+		if arg.err != nil {
+			readResponse.resultErr = arg.err
+			outBuf <- readResponse
+			continue
 		}
-		ok = true
 
-		r.metadataLookup[entry.Index] = metadata
-	}
+		arg.bytes.IncRef()
+		decoder.Reset(arg.bytes.Get())
+		entry, err := decoder.DecodeLogEntry()
+		if err != nil {
+			readResponse.resultErr = err
+			outBuf <- readResponse
+			continue
+		}
 
-	if !ok {
-		resultErr = errCommitLogReaderMissingLogMetadata
-		return
-	}
+		if len(entry.Metadata) != 0 {
+			metadataDecoder.Reset(entry.Metadata)
+			decoded, err := metadataDecoder.DecodeLogMetadata()
+			if err != nil {
+				readResponse.resultErr = err
+				outBuf <- readResponse
+				continue
+			}
 
-	series = metadata
-	datapoint = ts.Datapoint{
-		Timestamp: time.Unix(0, entry.Timestamp),
-		Value:     entry.Value,
+			id := r.bytesPool.Get(len(decoded.ID))
+			id.IncRef()
+			id.AppendAll(decoded.ID)
+
+			namespace := r.bytesPool.Get(len(decoded.Namespace))
+			namespace.IncRef()
+			namespace.AppendAll(decoded.Namespace)
+
+			r.metadata.Lock()
+			r.metadata.lookup[entry.Index] = Series{
+				UniqueIndex: entry.Index,
+				ID:          ts.BinaryID(id),
+				Namespace:   ts.BinaryID(namespace),
+				Shard:       decoded.Shard,
+			}
+			waitGroup, ok := r.metadata.wgs[entry.Index]
+			// Another goroutine is blocked waiting for this metadata
+			if ok {
+				waitGroup.Done()
+			}
+			r.metadata.Unlock()
+			namespace.DecRef()
+			id.DecRef()
+		}
+
+		var pendingResult *sync.WaitGroup
+		var metadata Series
+		var hasMetadata bool
+		var ok bool
+
+		// Try cheap read lock first
+		r.metadata.RLock()
+		metadata, hasMetadata = r.metadata.lookup[entry.Index]
+		r.metadata.RUnlock()
+
+		// The required metadata hasn't been processed yet
+		if !hasMetadata {
+			r.metadata.Lock()
+			metadata, hasMetadata = r.metadata.lookup[entry.Index]
+			if !hasMetadata {
+				waitGroup, hasWaitGroup := r.metadata.wgs[entry.Index]
+				if hasWaitGroup {
+					pendingResult = waitGroup
+				} else {
+					pendingResult = &sync.WaitGroup{}
+					pendingResult.Add(1)
+					r.metadata.wgs[entry.Index] = pendingResult
+				}
+			}
+			r.metadata.Unlock()
+		}
+
+		// The required metadata hasn't been processed yet
+		if pendingResult != nil {
+			pendingResult.Wait()
```

Comment:
This can cause an infinite wait if the input file is illformed and we don't receive a metadata entry for a series. We should guard against that case somehow (either timed or expect all metadata within a certain number of entries)

Reply:
i'd do the same as what you're doing in the `errCommitLogReaderMissingLogMetadata` case below